DOCSP-31470: describe connection pool in perf page (#93)
mongodb/docs-rust: source/fundamentals/performance.txt

   :language: rust
   :dedent:

.. _rust-performance-pool:

Connection Pool
---------------

Every ``Client`` instance has a built-in connection pool for each server
in your MongoDB topology. Connection pools open sockets on demand to
support concurrent requests to MongoDB in your application. You can
tune the connection pool to best fit the needs of your application
and optimize performance.

.. tip::
   
   To learn more about configuring a connection pool, see
   :manual:`Tuning Your Connection Pool Settings
   </tutorial/connection-pool-performance-tuning/>` in the Server manual.

The maximum size of each connection pool is set by the ``max_pool_size``
option, which defaults to ``10``. If the number of in-use connections to
a server reaches the value of ``max_pool_size``, the next request to

> that server will wait until a connection becomes available.

Reviewer: S: since the style guide likes present tense

```suggestion
that server waits until a connection becomes available.
```

In addition to the sockets needed to support your application's requests,
each ``Client`` instance opens two more sockets per server
in your MongoDB topology for monitoring the server's state.
For example, a client connected to a three-node replica set opens six
monitoring sockets. If the application uses the default setting for
``max_pool_size`` and only queries the primary (default) node, then
there can be at most 16 total connections in the connection pool. If the
application uses a read preference to query the secondary nodes, those
connection pools grow and there can be 36 total connections.

To support high numbers of concurrent MongoDB requests
within one process, you can increase the value of the ``max_pool_size``

> option. The following code demonstrates how to instantiate a ``Client``
> while specifying a value for ``max_pool_size``:

Reviewer: S: nitpicky suggestion, but since the focus is how to increase `max_pool_size`, I think this wording would be better:

```suggestion
option. The following code demonstrates how to specify a value for ``max_pool_size`` when instantiating a ``Client``:
```

.. code-block:: rust
   
   let mut client_options = ClientOptions::parse("<connection string>").await?;
   client_options.max_pool_size = Some(20);

   let client = Client::with_options(client_options)?;

Connection pools are rate-limited. The ``max_connecting`` option
determines the number of connections that the pool can create in
parallel at any time. For example, if the value of ``max_connecting`` is
``2``, as is by default, the third request that attempts to concurrently
check out a connection succeeds only when one of the following cases occurs:

> - The connection pool finishes creating a connection and there are fewer
>   than ``max_pool_size`` connections in the pool.

Reviewer: S: I think using the option's name as a stand-in for a value itself is a little confusing; you could reword to something like this:

```suggestion
- The connection pool finishes creating a connection and the number of connections in the pool is less than the value of ``max_pool_size``.
```

Reviewer: Applies elsewhere on the page (line 132)

- An existing connection is checked back into the pool.
- The driver's ability to reuse existing connections improves due to

>   rate-limits on connection creation.

Reviewer: S: I don't think the hyphen is necessary (unless rate-limit is usually written that way)

```suggestion
  rate limits on connection creation.
```

You can set the minimum number of concurrent connections to
each server with the ``min_pool_size`` option, which defaults to ``0``.
The driver initializes the connection pool with this number of sockets. If
sockets are closed and the total number of sockets, both in use and

> idle, to drop below the minimum, the connection pool opens more sockets until the

Reviewer: I:

```suggestion
idle, drops below the minimum, the connection pool opens more sockets until the
```

minimum is reached.

You can set the maximum amount of time that a connection can
remain idle in the pool by setting the ``max_idle_time`` option.
Once a connection has been idle for ``max_idle_time``, the connection

> pool removes and replaces it. This option defaults to ``0``, or no limit.

Reviewer: S: just to clarify "it"

```suggestion
pool removes and replaces that connection. This option defaults to ``0``, or no limit.
```

each ``Client`` instance opens two more sockets per server
in your MongoDB topology for monitoring the server's state.
For example, a client connected to a three-node replica set opens six
monitoring sockets. If the application uses the default setting for
``max_pool_size`` and only queries the primary (default) node, then
there can be at most 16 total connections in the connection pool. If the
application uses a read preference to query the secondary nodes, those
connection pools grow and there can be 36 total connections.

To support high numbers of concurrent MongoDB requests
within one process, you can increase the value of the ``max_pool_size``
option. The following code demonstrates how to specify a value for
``max_pool_size`` when instantiating a ``Client``:

.. code-block:: rust
   
   let mut client_options = ClientOptions::parse("<connection string>").await?;
   client_options.max_pool_size = Some(20);

   let client = Client::with_options(client_options)?;

Connection pools are rate-limited. The ``max_connecting`` option
determines the number of connections that the pool can create in
parallel at any time. For example, if the value of ``max_connecting`` is

> ``2``, as is by default, the third request that attempts to concurrently

Reviewer: S: the wording of "as is by default" seems a little clunky, could be:

```suggestion
``2``, its default value, the third request that attempts to concurrently
```

check out a connection succeeds only when one of the following cases occurs:

- The connection pool finishes creating a connection and the number of

>   connections in the pool is less than the value of ``max_pool_size``.

Reviewer 2: ```suggestion
  connections in the pool is less than or equal to the value of ``max_pool_size``.
```
